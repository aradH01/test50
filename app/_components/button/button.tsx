"use client";

import { ButtonHTMLAttributes, ReactNode } from "react";
import { cn } from "@lib/utils";

export type ButtonVariant =
  | "default"
  | "destructive"
  | "outline"
  | "secondary"
  | "ghost"
  | "link";
export type ButtonSize = "default" | "sm" | "lg" | "icon";

const getVariantClasses = (variant: ButtonVariant): string => {
  const variants = {
    default: "bg-blue-600 text-white hover:bg-blue-700 active:bg-blue-800",
    destructive: "bg-red-600 text-white hover:bg-red-700 active:bg-red-800",
    outline:
      "border border-gray-300 bg-white text-gray-700 hover:bg-gray-50 active:bg-gray-100",
    secondary: "bg-gray-600 text-white hover:bg-gray-700 active:bg-gray-800",
    ghost: "text-gray-700 hover:bg-gray-100 active:bg-gray-200",
    link: "text-blue-600 underline-offset-4 hover:underline active:text-blue-800",
  };
  return variants[variant];
};

const getSizeClasses = (size: ButtonSize): string => {
  const sizes = {
    default: "h-10 px-4 py-2",
    sm: "h-9 rounded-md px-3 text-xs",
    lg: "h-11 rounded-md px-8 text-base",
    icon: "h-10 w-10",
  };
  return sizes[size];
};

export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  children: ReactNode;
  variant?: ButtonVariant;
  size?: ButtonSize;
  loading?: boolean;
};

export function Button({
  className,
  variant = "default",
  size = "default",
  children,
  loading = false,
  disabled,
  ...props
}: ButtonProps) {
  const baseClasses =
    "inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-blue-500 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50";

  return (
    <button
      className={cn(
        baseClasses,
        getVariantClasses(variant),
        getSizeClasses(size),
        className,
      )}
      disabled={disabled || loading}
      {...props}
    >
      {loading && (
        <svg
          className="mr-2 h-4 w-4 animate-spin"
          xmlns="http://www.w3.org/2000/svg"
          fill="none"
          viewBox="0 0 24 24"
        >
          <circle
            className="opacity-25"
            cx="12"
            cy="12"
            r="10"
            stroke="currentColor"
            strokeWidth="4"
          />
          <path
            className="opacity-75"
            fill="currentColor"
            d="m4 12a8 8 0 0 1 8-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 0 1 4 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
          />
        </svg>
      )}
      {children}
    </button>
  );
}
