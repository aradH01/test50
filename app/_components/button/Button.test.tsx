/// <reference types="jest" />
/// <reference types="@testing-library/jest-dom" />

import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Button } from "./button";

describe("Button", () => {
  // Arrange, Act, Assert pattern

  describe("Basic functionality", () => {
    it("renders button with text", () => {
      // Arrange & Act
      render(<Button>Click me</Button>);

      // Assert
      expect(
        screen.getByRole("button", { name: "Click me" }),
      ).toBeInTheDocument();
    });

    it("handles click events", async () => {
      // Arrange
      const handleClick = jest.fn();
      const user = userEvent.setup();
      render(<Button onClick={handleClick}>Click me</Button>);

      // Act
      await user.click(screen.getByRole("button"));

      // Assert
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it("accepts custom className", () => {
      // Arrange & Act
      render(<Button className="custom-class">Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("custom-class");
    });
  });

  describe("Variants", () => {
    it("applies default variant classes", () => {
      // Arrange & Act
      render(<Button>Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("bg-blue-600");
    });

    it("applies destructive variant classes", () => {
      // Arrange & Act
      render(<Button variant="destructive">Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("bg-red-600");
    });

    it("applies outline variant classes", () => {
      // Arrange & Act
      render(<Button variant="outline">Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("border-gray-300");
    });

    it("applies secondary variant classes", () => {
      // Arrange & Act
      render(<Button variant="secondary">Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("bg-gray-600");
    });

    it("applies ghost variant classes", () => {
      // Arrange & Act
      render(<Button variant="ghost">Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("text-gray-700");
    });

    it("applies link variant classes", () => {
      // Arrange & Act
      render(<Button variant="link">Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("text-blue-600");
    });
  });

  describe("Sizes", () => {
    it("applies default size classes", () => {
      // Arrange & Act
      render(<Button>Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("h-10");
    });

    it("applies small size classes", () => {
      // Arrange & Act
      render(<Button size="sm">Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("h-9");
    });

    it("applies large size classes", () => {
      // Arrange & Act
      render(<Button size="lg">Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("h-11");
    });

    it("applies icon size classes", () => {
      // Arrange & Act
      render(<Button size="icon">ğŸš€</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("h-10", "w-10");
    });
  });

  describe("Loading state", () => {
    it("disables button when loading", () => {
      // Arrange & Act
      render(<Button loading>Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toBeDisabled();
    });

    it("shows loading spinner when loading", () => {
      // Arrange & Act
      render(<Button loading>Button</Button>);

      // Assert
      expect(
        screen.getByRole("button").querySelector("svg"),
      ).toBeInTheDocument();
    });

    it("prevents click when loading", async () => {
      // Arrange
      const handleClick = jest.fn();
      const user = userEvent.setup();
      render(
        <Button loading onClick={handleClick}>
          Button
        </Button>,
      );

      // Act
      await user.click(screen.getByRole("button"));

      // Assert
      expect(handleClick).not.toHaveBeenCalled();
    });
  });

  describe("Disabled state", () => {
    it("disables button when disabled prop is true", () => {
      // Arrange & Act
      render(<Button disabled>Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toBeDisabled();
    });

    it("prevents click when disabled", async () => {
      // Arrange
      const handleClick = jest.fn();
      const user = userEvent.setup();
      render(
        <Button disabled onClick={handleClick}>
          Button
        </Button>,
      );

      // Act
      await user.click(screen.getByRole("button"));

      // Assert
      expect(handleClick).not.toHaveBeenCalled();
    });

    it("applies disabled styling", () => {
      // Arrange & Act
      render(<Button disabled>Button</Button>);

      // Assert
      expect(screen.getByRole("button")).toHaveClass("disabled:opacity-50");
    });
  });

  describe("HTML attributes", () => {
    it("forwards HTML button attributes", () => {
      // Arrange & Act
      render(
        <Button type="submit" form="test-form" data-testid="test-button">
          Submit
        </Button>,
      );

      // Assert
      const button = screen.getByRole("button");
      expect(button).toHaveAttribute("type", "submit");
      expect(button).toHaveAttribute("form", "test-form");
      expect(button).toHaveAttribute("data-testid", "test-button");
    });
  });

  // Snapshot test
  it("matches snapshot", () => {
    // Arrange & Act
    const { container } = render(<Button>Test Button</Button>);

    // Assert
    expect(container.firstChild).toMatchSnapshot();
  });

  it("matches loading state snapshot", () => {
    // Arrange & Act
    const { container } = render(<Button loading>Loading Button</Button>);

    // Assert
    expect(container.firstChild).toMatchSnapshot();
  });
});
