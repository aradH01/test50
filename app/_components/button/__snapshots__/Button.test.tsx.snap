// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button matches loading state snapshot 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-blue-500 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-blue-600 text-white hover:bg-blue-700 active:bg-blue-800 h-10 px-4 py-2"
  disabled=""
>
  <svg
    class="mr-2 h-4 w-4 animate-spin"
    fill="none"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="opacity-25"
      cx="12"
      cy="12"
      r="10"
      stroke="currentColor"
      stroke-width="4"
    />
    <path
      class="opacity-75"
      d="m4 12a8 8 0 0 1 8-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 0 1 4 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
      fill="currentColor"
    />
  </svg>
  Loading Button
</button>
`;

exports[`Button matches snapshot 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-blue-500 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-blue-600 text-white hover:bg-blue-700 active:bg-blue-800 h-10 px-4 py-2"
>
  Test Button
</button>
`;
